File: laboratory-classes/lab-list5.ml
(* Jakub Czerwinski *)

(* lista heterogeniczna *)
type 'a ticket = Covered of (unit -> 'a) | Uncovered of 'a;;

(* Zadanie 1 *)
let rec buyTicket table number =
  match (table, number) with
    | ((Covered h)::t, 1) -> Uncovered(h()) :: t
    | (h::t, n) ->  h :: buyTicket t (n - 1)
    | ([], _) -> [];;

(* Zakladamy ze podana liczba jest z przedzialu [1; table.length] *)
let table = [Covered(fun() -> 1); Covered(fun() -> 2); Covered(fun() -> 3); Covered(fun() -> 4); Covered(fun() -> 5)];;
let tableString = [Covered(fun() -> "Talon"); Covered(fun() -> "na"); Covered(fun() -> "balon"); Covered(fun() -> "!")];;

let t3 = buyTicket table 3;;
let t5 = buyTicket t3 5;;

buyTicket [] 3;;
buyTicket tableString 4;;

File: laboratory-classes/lab-list5.scala
// Jakub Czerwinski

// Zadanie 2
def subListL[A] (lxs: LazyList[A]) : LazyList[List[A]] =
  def helper(lys: LazyList[A], n: Int) : LazyList[List[A]] =
    (lys, n) match
      case (LazyList(), _) => LazyList(List())
      case (lh #:: lt, n) => lxs.take(n).toList #:: helper(lt, n + 1)
  helper(lxs, 1);

subListL(LazyList.from(1)).take(3).toList
subListL(LazyList.from(5)).take(5).toList
subListL(LazyList('k', 'u', 'b', 'a')).take(4).toList
subListL(LazyList()).take(3).toList
